{
    "url": "https://api.github.com/repos/jax-ml/jax/issues/26112",
    "repository_url": "https://api.github.com/repos/jax-ml/jax",
    "labels_url": "https://api.github.com/repos/jax-ml/jax/issues/26112/labels{/name}",
    "comments_url": "https://api.github.com/repos/jax-ml/jax/issues/26112/comments",
    "events_url": "https://api.github.com/repos/jax-ml/jax/issues/26112/events",
    "html_url": "https://github.com/jax-ml/jax/issues/26112",
    "id": 2811839921,
    "node_id": "I_kwDOCTkjjc6nmUWx",
    "number": 26112,
    "title": "linalg.norm doesn't support more than 2 explicit axes, confusing error message",
    "user": {
      "login": "Sohl-Dickstein",
      "id": 498544,
      "node_id": "MDQ6VXNlcjQ5ODU0NA==",
      "avatar_url": "https://avatars.githubusercontent.com/u/498544?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/Sohl-Dickstein",
      "html_url": "https://github.com/Sohl-Dickstein",
      "followers_url": "https://api.github.com/users/Sohl-Dickstein/followers",
      "following_url": "https://api.github.com/users/Sohl-Dickstein/following{/other_user}",
      "gists_url": "https://api.github.com/users/Sohl-Dickstein/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/Sohl-Dickstein/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/Sohl-Dickstein/subscriptions",
      "organizations_url": "https://api.github.com/users/Sohl-Dickstein/orgs",
      "repos_url": "https://api.github.com/users/Sohl-Dickstein/repos",
      "events_url": "https://api.github.com/users/Sohl-Dickstein/events{/privacy}",
      "received_events_url": "https://api.github.com/users/Sohl-Dickstein/received_events",
      "type": "User",
      "user_view_type": "public",
      "site_admin": false
    },
    "labels": [
      {
        "id": 1104782284,
        "node_id": "MDU6TGFiZWwxMTA0NzgyMjg0",
        "url": "https://api.github.com/repos/jax-ml/jax/labels/bug",
        "name": "bug",
        "color": "d73a4a",
        "default": true,
        "description": "Something isn't working"
      }
    ],
    "state": "closed",
    "locked": false,
    "assignee": {
      "login": "jakevdp",
      "id": 781659,
      "node_id": "MDQ6VXNlcjc4MTY1OQ==",
      "avatar_url": "https://avatars.githubusercontent.com/u/781659?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/jakevdp",
      "html_url": "https://github.com/jakevdp",
      "followers_url": "https://api.github.com/users/jakevdp/followers",
      "following_url": "https://api.github.com/users/jakevdp/following{/other_user}",
      "gists_url": "https://api.github.com/users/jakevdp/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/jakevdp/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/jakevdp/subscriptions",
      "organizations_url": "https://api.github.com/users/jakevdp/orgs",
      "repos_url": "https://api.github.com/users/jakevdp/repos",
      "events_url": "https://api.github.com/users/jakevdp/events{/privacy}",
      "received_events_url": "https://api.github.com/users/jakevdp/received_events",
      "type": "User",
      "user_view_type": "public",
      "site_admin": false
    },
    "assignees": [
      {
        "login": "jakevdp",
        "id": 781659,
        "node_id": "MDQ6VXNlcjc4MTY1OQ==",
        "avatar_url": "https://avatars.githubusercontent.com/u/781659?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/jakevdp",
        "html_url": "https://github.com/jakevdp",
        "followers_url": "https://api.github.com/users/jakevdp/followers",
        "following_url": "https://api.github.com/users/jakevdp/following{/other_user}",
        "gists_url": "https://api.github.com/users/jakevdp/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/jakevdp/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/jakevdp/subscriptions",
        "organizations_url": "https://api.github.com/users/jakevdp/orgs",
        "repos_url": "https://api.github.com/users/jakevdp/repos",
        "events_url": "https://api.github.com/users/jakevdp/events{/privacy}",
        "received_events_url": "https://api.github.com/users/jakevdp/received_events",
        "type": "User",
        "user_view_type": "public",
        "site_admin": false
      }
    ],
    "milestone": null,
    "comments": 3,
    "created_at": "2025-01-26T23:18:56Z",
    "updated_at": "2025-01-27T19:25:07Z",
    "closed_at": "2025-01-27T19:25:06Z",
    "author_association": "CONTRIBUTOR",
    "sub_issues_summary": {
      "total": 0,
      "completed": 0,
      "percent_completed": 0
    },
    "active_lock_reason": null,
    "body": "### Description\n\nUnsure if this is a bug or a feature request. The snippet:\n```\na = jnp.ones((1, 1, 1,))\nprint(f\"{jnp.linalg.norm(a, axis=(0, 1, 2,))=}\")\n```\nthrows an error:\n```\nValueError: Invalid axis values ((0, 1, 2)) for jnp.linalg.norm.\n```\n\nThe same operation in numpy (rather than jax.numpy) returns the more interpretable error message:\n```\nValueError: Improper number of dimensions to norm.\n```\n\nI think linalg.norm should accept any number of dimensions in the axis argument! If the current behavior is retained, the error message should be modified to explain why the axis values are invalid (because linalg.norm only accepts up to two axis values).\n\n### System info (python version, jaxlib version, accelerator, etc.)\n\nnot relevant",
    "closed_by": {
      "login": "copybara-service[bot]",
      "id": 56741989,
      "node_id": "MDM6Qm90NTY3NDE5ODk=",
      "avatar_url": "https://avatars.githubusercontent.com/in/44061?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/copybara-service%5Bbot%5D",
      "html_url": "https://github.com/apps/copybara-service",
      "followers_url": "https://api.github.com/users/copybara-service%5Bbot%5D/followers",
      "following_url": "https://api.github.com/users/copybara-service%5Bbot%5D/following{/other_user}",
      "gists_url": "https://api.github.com/users/copybara-service%5Bbot%5D/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/copybara-service%5Bbot%5D/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/copybara-service%5Bbot%5D/subscriptions",
      "organizations_url": "https://api.github.com/users/copybara-service%5Bbot%5D/orgs",
      "repos_url": "https://api.github.com/users/copybara-service%5Bbot%5D/repos",
      "events_url": "https://api.github.com/users/copybara-service%5Bbot%5D/events{/privacy}",
      "received_events_url": "https://api.github.com/users/copybara-service%5Bbot%5D/received_events",
      "type": "Bot",
      "user_view_type": "public",
      "site_admin": false
    },
    "reactions": {
      "url": "https://api.github.com/repos/jax-ml/jax/issues/26112/reactions",
      "total_count": 0,
      "+1": 0,
      "-1": 0,
      "laugh": 0,
      "hooray": 0,
      "confused": 0,
      "heart": 0,
      "rocket": 0,
      "eyes": 0
    },
    "timeline_url": "https://api.github.com/repos/jax-ml/jax/issues/26112/timeline",
    "performed_via_github_app": null,
    "state_reason": "completed",
    "comments_url_body": [
      {
        "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2614681400",
        "html_url": "https://github.com/jax-ml/jax/issues/26112#issuecomment-2614681400",
        "issue_url": "https://api.github.com/repos/jax-ml/jax/issues/26112",
        "id": 2614681400,
        "node_id": "IC_kwDOCTkjjc6b2N84",
        "user": {
          "login": "jakevdp",
          "id": 781659,
          "node_id": "MDQ6VXNlcjc4MTY1OQ==",
          "avatar_url": "https://avatars.githubusercontent.com/u/781659?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/jakevdp",
          "html_url": "https://github.com/jakevdp",
          "followers_url": "https://api.github.com/users/jakevdp/followers",
          "following_url": "https://api.github.com/users/jakevdp/following{/other_user}",
          "gists_url": "https://api.github.com/users/jakevdp/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/jakevdp/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/jakevdp/subscriptions",
          "organizations_url": "https://api.github.com/users/jakevdp/orgs",
          "repos_url": "https://api.github.com/users/jakevdp/repos",
          "events_url": "https://api.github.com/users/jakevdp/events{/privacy}",
          "received_events_url": "https://api.github.com/users/jakevdp/received_events",
          "type": "User",
          "user_view_type": "public",
          "site_admin": false
        },
        "created_at": "2025-01-27T01:04:23Z",
        "updated_at": "2025-01-27T01:04:23Z",
        "author_association": "COLLABORATOR",
        "body": "Thanks for the report! JAX follows the NumPy API convention for the meaning of `axis` in `norm`, and in NumPy 3 axes leads to an error. I suspect the reason is that the norm along one axis is treated as a vector norm, the norm along two axes is treated as a matrix norm, and there's no good convention for a norm of a 3D tensor. Still, we could probably improve the error message \u2013 what do you think?",
        "reactions": {
          "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2614681400/reactions",
          "total_count": 0,
          "+1": 0,
          "-1": 0,
          "laugh": 0,
          "hooray": 0,
          "confused": 0,
          "heart": 0,
          "rocket": 0,
          "eyes": 0
        },
        "performed_via_github_app": null
      },
      {
        "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2616408494",
        "html_url": "https://github.com/jax-ml/jax/issues/26112#issuecomment-2616408494",
        "issue_url": "https://api.github.com/repos/jax-ml/jax/issues/26112",
        "id": 2616408494,
        "node_id": "IC_kwDOCTkjjc6b8zmu",
        "user": {
          "login": "Sohl-Dickstein",
          "id": 498544,
          "node_id": "MDQ6VXNlcjQ5ODU0NA==",
          "avatar_url": "https://avatars.githubusercontent.com/u/498544?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/Sohl-Dickstein",
          "html_url": "https://github.com/Sohl-Dickstein",
          "followers_url": "https://api.github.com/users/Sohl-Dickstein/followers",
          "following_url": "https://api.github.com/users/Sohl-Dickstein/following{/other_user}",
          "gists_url": "https://api.github.com/users/Sohl-Dickstein/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/Sohl-Dickstein/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/Sohl-Dickstein/subscriptions",
          "organizations_url": "https://api.github.com/users/Sohl-Dickstein/orgs",
          "repos_url": "https://api.github.com/users/Sohl-Dickstein/repos",
          "events_url": "https://api.github.com/users/Sohl-Dickstein/events{/privacy}",
          "received_events_url": "https://api.github.com/users/Sohl-Dickstein/received_events",
          "type": "User",
          "user_view_type": "public",
          "site_admin": false
        },
        "created_at": "2025-01-27T17:14:32Z",
        "updated_at": "2025-01-27T17:14:32Z",
        "author_association": "CONTRIBUTOR",
        "body": "Definitely a more descriptive error message would have saved me significant\r\ndebugging! A more descriptive message would be appreciated and will help\r\nothers in the future.\r\n\r\nI think breaking compatibility so that jnp works when np errors out would\r\nbe pretty defensible, but I understand the argument against.\r\n\r\nI think there is an obvious extension of norms to arbitrary dimensions,\r\nwhich np.linalg.norm actually uses. Numpy returns the expected thing if you\r\ncall limalg.norm with a 3d array *without* specifying indices, and only\r\nthrows an error if you specify those same three indices explicitly.\r\n\r\nOn Sun, Jan 26, 2025, 17:04 Jake Vanderplas ***@***.***>\r\nwrote:\r\n\r\n> Thanks for the report! JAX follows the NumPy API convention for the\r\n> meaning of axis in norm, and in NumPy 3 axes leads to an error. I suspect\r\n> the reason is that the norm along one axis is treated as a vector norm, the\r\n> norm along two axes is treated as a matrix norm, and there's no good\r\n> convention for a norm of a 3D tensor. Still, we could probably improve the\r\n> error message \u2013 what do you think?\r\n>\r\n> \u2014\r\n> Reply to this email directly, view it on GitHub\r\n> <https://github.com/jax-ml/jax/issues/26112#issuecomment-2614681400>, or\r\n> unsubscribe\r\n> <https://github.com/notifications/unsubscribe-auth/AADZW4CHZ3Z45WM6XH5VRDT2MWA2ZAVCNFSM6AAAAABV45HBJSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJUGY4DCNBQGA>\r\n> .\r\n> You are receiving this because you authored the thread.Message ID:\r\n> ***@***.***>\r\n>\r\n",
        "reactions": {
          "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2616408494/reactions",
          "total_count": 0,
          "+1": 0,
          "-1": 0,
          "laugh": 0,
          "hooray": 0,
          "confused": 0,
          "heart": 0,
          "rocket": 0,
          "eyes": 0
        },
        "performed_via_github_app": null
      },
      {
        "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2616503708",
        "html_url": "https://github.com/jax-ml/jax/issues/26112#issuecomment-2616503708",
        "issue_url": "https://api.github.com/repos/jax-ml/jax/issues/26112",
        "id": 2616503708,
        "node_id": "IC_kwDOCTkjjc6b9K2c",
        "user": {
          "login": "jakevdp",
          "id": 781659,
          "node_id": "MDQ6VXNlcjc4MTY1OQ==",
          "avatar_url": "https://avatars.githubusercontent.com/u/781659?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/jakevdp",
          "html_url": "https://github.com/jakevdp",
          "followers_url": "https://api.github.com/users/jakevdp/followers",
          "following_url": "https://api.github.com/users/jakevdp/following{/other_user}",
          "gists_url": "https://api.github.com/users/jakevdp/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/jakevdp/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/jakevdp/subscriptions",
          "organizations_url": "https://api.github.com/users/jakevdp/orgs",
          "repos_url": "https://api.github.com/users/jakevdp/repos",
          "events_url": "https://api.github.com/users/jakevdp/events{/privacy}",
          "received_events_url": "https://api.github.com/users/jakevdp/received_events",
          "type": "User",
          "user_view_type": "public",
          "site_admin": false
        },
        "created_at": "2025-01-27T17:50:27Z",
        "updated_at": "2025-01-27T17:50:27Z",
        "author_association": "COLLABORATOR",
        "body": "> I think there is an obvious extension of norms to arbitrary dimensions, which np.linalg.norm actually uses.\n\nNumPy only uses this extension when `ord=None`, which effectively becomes a generalized 2-norm; see https://github.com/numpy/numpy/issues/14215 for a past discussion of this and NumPy's rationale for not changing this undocumented behavior.\n\nI think we could probably improve the documentation here, and also improve the error for unsupported axes patterns.",
        "reactions": {
          "url": "https://api.github.com/repos/jax-ml/jax/issues/comments/2616503708/reactions",
          "total_count": 0,
          "+1": 0,
          "-1": 0,
          "laugh": 0,
          "hooray": 0,
          "confused": 0,
          "heart": 0,
          "rocket": 0,
          "eyes": 0
        },
        "performed_via_github_app": null
      }
    ]
  }